Upgrade kernels to mitigate Stack Clash

conorsch opened the pull request:
Implements the mitigations described in https://github.com/freedomofpress/securedrop/issues/1861. These changes are SecureDrop-specific, and in addition to updating the grsecurity kernel patch version, also provide a sysctl modification to `vm.heap_stack_gap`, raising it 64KB -> 1MB, as [recommended by grsecurity](https://grsecurity.net/an_ancient_kernel_hole_is_not_closed.php). The sysctl options are rolled into the `securedrop-grsec` metapackage, same as the paxctl flags added via #91.

Also commits the latest kernel config used for SecureDrop, which is the default Ubuntu Trusty config with the grsecurity patch set applied with high security settings, as [indicated in the docs](https://github.com/freedomofpress/ansible-role-grsecurity/blob/a62731bf3c30ed6aab491e892693ce96ed501d87/docs/securedrop_metapackage.md), and thereby closes #3. We'll need to update that if and when we move to the 4.x kernel series.

conorsch committed:
Adds Stefan Bader (Canonical) GPG key

Another trusted kernel maintainer at Canonical. Pulling in the new key
along with the others, since it's not predictable which key was used to
sign the latest commit on the ubuntu-overlay.

conorsch committed:
Commits latest kernel config used for SecureDrop

Allows us to publish the current kernel config for Ubuntu Trusty, used
for SecureDrop, for transparency and accountability. The config is
extremely simple: default Ubuntu Trusty settings, with the grsecurity
high-security options enabled on top.

Updates the Ansible playbook for the SecureDrop kernel build process to
import the new config automatically. This will still need to change
if/when we migrate to the 4.x kernel series.

conorsch committed:
Provides sysctl configuration in securedrop-grsec metapackage

Creates sysctl options file in metapackage, and drops it into
/etc/sysctl.d/. The strategy here is to provide unattended upgrade
capability for tweaking sysctl options, as a defense-in-depth mitigation
for the Stack Clash vuln. See here for more info:

https://grsecurity.net/an_ancient_kernel_hole_is_not_closed.php

conorsch committed:
Bumps metapackage version number with suffix

Using the "+r1" suffix appended to the version for the
`securedrop-grsec` metapackage because otherwise we won't have a unique
Version field, meaning automatic updates won't install the new package.
The previous Version field was simply "3.14.79", the same as the Linux
source version, and since we're reusing that source (but with a newer
grsecurity patch), we need a unique identifier.

conorsch committed:
Bumps revision in SecureDrop kernel version

The previous kernel, currently in use by production SecureDrop
instances, shared the same kernel source version, 3.14.79. The default
revision for manual compiles supplied by `make-kpkg` is "10.00.Custom",
according to the make-kpkg man page, but the kernel version is prepended
automatically if you don't use the `--revision` flag, so it's actually
"3.14.79-grsec-10.00.Custom".

In order to ensure a smooth upgrade flow, we need to provide semver
compatibility with the previous version. Using "11.00.Custom" suffices.
Given the lack of automatic prepending when the `--revision` flag is
used, we must specify the entire kernel version as a prefix, i.e.
"3.14.79-grsec-11.00.Custom". Then we get sane version parsing according
to Debian package guidelines.

See the Debian Policy Guide for more info on how the version field is
parsed [0].

[0] https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version

conorsch committed:
Enforces specific build targets for SecureDrop

Historically we've been maintaining both kernel image and header
packages, and posting them to the FPF apt repo (apt.freedom.press),
for use with SecureDrop. Only the image is technically necessary, but to
maintain backwards compatibility, let's ensure that the headers are
built as part of the SecureDrop-specific build playbook.

In addition, we must maintain the source packages, although at this time
we are not providing those publicly in the apt repository. Offers of source
for the hosted binaries will be available in the SecureDrop project
repo.

msheiny commented on roles/build-grsec-kernel/tasks/gpg_keys.yml:
```diff
@@ -25,3 +25,9 @@
   when: grsecurity_build_include_ubuntu_overlay == true
+
+- name: Import Stefan Bader GPG key (Canonical/Ubuntu Kernel Team)
+  command: gpg --keyserver {{ grsecurity_build_gpg_keyserver }} --recv-keys E8675DEECBEECEA3
```
Thinking these keys should be a variable

psivesely replied:
Prefer the full fingerprint `DB5D7CCAF3994E3395DA4D3EE8675DEECBEECEA3`. While expensive, long key ID collisions are possible.

msheiny replied:
Hey @conorsch - I think what @fowlslegs is requesting is a good call here (plus convert that to a default var)-- after that's done im 👍 up for margin' here

msheiny commented on roles/build-grsec-metapackage/files/securedrop-sysctl-options.conf:
```diff
@@ -0,0 +1 @@
+vm.heap_stack_gap=1048576
```
I think should also be a default variable / template

psivesely reviewed (dismissed):
The good:
* Correct permissions and ownerships.
* Correct sysctl parameters and use of `sysctl.d` file.
* Grsec configuration options listed in https://github.com/freedomofpress/securedrop/issues/1861#issuecomment-309636077 are enabled.
* Building headers too seems smart.
* Versioning seems in line with Debian packaging standards and upgrades should happen as expected.

The bad:
* Incorrect use of GPG.

msheiny commented:
@conorsch - i really want to do a full test against this since its been a while since i ran this role -- im assuming this PR isnt urgent to get in since youve already built it but if my assumption is incorrect let me know

conorsch commented:
@msheiny That's right, changes are already live and shipped. In order to test locally, I recommend something like:

```
vagrant up --no-provision grsec-build-securedrop
ansible-playbook -vv --diff -i .vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory examples/build-grsecurity-kernel-securedrop.yml  --tags kernel
```

The manual playbook invocation is necessary for the `prompt` tasks to permit input from stdin, otherwise they'll be skipped and the role will fail.

msheiny committed:
Allow settings kernel-img.conf values from vars

This is an optional variable that allows setting up kernel-img.conf properties.
Some of these values in particular were needed during install - specifically,
`clobber_modules` that was the catalyst for adding this option. If the user
doesnt modify this value it will continue to utilize the default conf that is
dropped upon first install.

msheiny commented:
So two issues I ran into while deploying -- not really code specific but worth noting for the future:

* When you try to deploy the same kernel version from deb package directly -- you'll get a big scary error about existing modules in place. The only way I found to get past this was to either backup/move the existing dir or set `clobber_modules` via `/etc/kernel-img.conf`. I added commit that allows one to do that via the second option.
* When we deploy newer versions of kernels to over-write existing kernels with the same version... this is problematic because it makes it harder to check to see whats running during operations. Sure - I can check the dpkg version of whats installed but what if the system hasnt been rebooted yet and is still running an older version? I think we should always update kernel version that displays when you query from within the os. in other words,  `uname -r` output should always be incrementing when we update kernels internally.

conorsch committed:
Converts metapackage sysctl options file to template

Allows site-level overrides to add additional sysctl options without
modifying the role.

conorsch committed:
Uses full-length fingerprints for GPG keys

Intended to avoid collision attacks against the short keyids.

conorsch committed:
Moves all GPG key fingerprints to default role vars

Provides a single point of override, so that additions or deletions to
the list of required signing keys for verification can be provided at
the site level, rather than requiring changes to the role.

conorsch commented:
@msheiny Revised GPG fingerprints to use longids, and moved the sysctl options to a vars-based template. Ready for re-review.

msheiny commented on roles/build-grsec-kernel/tasks/gpg_keys.yml:
```diff
@@ -1,27 +1,13 @@
+  command: gpg --keyserver {{ grsecurity_build_gpg_keyserver }} --recv-keys "{{ item.fingerprint }}"
+  register: gpg_import_linux_pubkeys_result
+  changed_when: "'imported: 1' in gpg_import_linux_pubkeys_result.item.stderr"
+  with_items: "{{ grsecurity_build_gpg_keys }}"
```
`with_items`  💟 . I'm a sucker for a good ansible yaml loop

msheiny commented on roles/build-grsec-kernel/defaults/main.yml:
```diff
@@ -70,3 +70,22 @@ grsecurity_build_fetch_packages_dest: "./"
+# List of GPG keys required for building grsecurity-patched kernel.
+grsecurity_build_gpg_keys:
+  - name: Greg Kroah-Hartman GPG key (Linux stable release signing key)
+    fingerprint: 647F28654894E3BD457199BE38DBBDC86092693E
```
Mmmmmmmmmmmmm fingerprinty

msheiny reviewed (approved):
i'll admit I didn't try to run this through this round - and only looked at the diff from last time. Thanks for addressing those fingerprint variable concerns 👍
